### PR TITLE
Fuzzy match

### PR DESCRIPTION
**Pull request by narekgharibyan**
- implemented fuzzy matching
 - added Dictionary::GetFuzzy() method
 - fixed use of .clang-format file
 - added ConstructorInitializerIndentWidth: 4 to .clang-format

Note: At the moment Python bindings are missing, will add them in a separate PR.

### CODE REVIEW

File: keyvi/check-style.sh [hunk 1]
```diff
@@ -8,7 +8,7 @@ cpplint_files=()
 for file in ${infiles}; do
   fqfile=`git rev-parse --show-toplevel`/${file}
   echo "Checking: ${file}"
-  if ! cmp -s ${fqfile} <(clang-format ${fqfile}); then
+  if ! cmp -s ${fqfile} <(clang-format -style=file ${fqfile}); then
```
Comment (hendrikmuhs) [inline keyvi/check-style.sh]: can you explain, is that something that does not work on your setup?

File: CMakeLists.txt [hunk 2]
```diff
@@ -62,7 +62,7 @@ FILE(GLOB_RECURSE UNIT_TEST_SOURCES RELATIVE ${CMAKE_CURRENT_SOURCE_DIR} keyvi/t
 add_executable(keyvicompiler keyvi/bin/keyvicompiler/keyvicompiler.cpp)
 add_executable(keyviinspector keyvi/bin/keyviinspector/keyviinspector.cpp)
 add_executable(keyvimerger keyvi/bin/keyvimerger/keyvimerger.cpp)
-add_executable(unit_test_all ${UNIT_TEST_SOURCES})
+add_executable(unit_test_all ${UNIT_TEST_SOURCES} keyvi/include/keyvi/dictionary/fuzzy_match.h)
```
Comment (hendrikmuhs) [inline CMakeLists.txt]: this should not be required

File: keyvi/check-style.sh (hunk 1 above)
Comment (narekgharibyan) [reply keyvi/check-style.sh]: At some point `clang-format` was not running, and it seemed to me that this had fixed it.
But just tested, it's fine even without it. Thx!

File: CMakeLists.txt (hunk 2 above)
Comment (narekgharibyan) [reply CMakeLists.txt]: done

File: keyvi/include/keyvi/dictionary/dictionary.h [hunk 3]
```diff
@@ -32,6 +32,7 @@
 #include "dictionary/fsa/automata.h"
 #include "dictionary/fsa/state_traverser.h"
 #include "dictionary/fsa/traverser_types.h"
+#include "dictionary/fuzzy_match.h"
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/dictionary.h]: I find it a bit confusing: 

- fuzzy_match.h sounds like a derived version of match.h, but that isn't the case.
- it is place top-level, do you think it will be used directly or always via dictionary?

File: keyvi/include/keyvi/dictionary/dictionary.h [hunk 4]
```diff
@@ -381,6 +382,12 @@ class Dictionary final {
     return MatchIterator::MakeIteratorPair(tfunc);
   }
 
+  MatchIterator::MatchIteratorPair GetFuzzy(const std::string& query, size_t max_edit_distance) {
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/dictionary.h]: could that be const?

File: keyvi/include/keyvi/dictionary/fuzzy_match.h [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+#include "dictionary/match.h"
+#include "dictionary/util/utf8_utils.h"
+
+namespace keyvi {
+namespace dictionary {
+
+class FuzzyMatch {
+ public:
+  FuzzyMatch(fsa::automata_t fsa, const std::string& query, size_t max_edit_distance)
+      : fsa_(std::move(fsa)), query_(query), max_edit_distance_(max_edit_distance) {
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/fuzzy_match.h]: I think std::move is not effective here

File: keyvi/include/keyvi/dictionary/fuzzy_match.h [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+#include "dictionary/fsa/automata.h"
+#include "dictionary/fsa/codepoint_state_traverser.h"
+#include "dictionary/fsa/traverser_types.h"
+#include "dictionary/match.h"
+#include "dictionary/util/utf8_utils.h"
+
+namespace keyvi {
+namespace dictionary {
+
+class FuzzyMatch {
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/fuzzy_match.h]: final?

(at least until we find a reason to derive something from it)

File: keyvi/include/keyvi/dictionary/fuzzy_match.h (hunk 5 above)
Comment (hendrikmuhs) [reply keyvi/include/keyvi/dictionary/fuzzy_match.h]: I would also init first_match_ as part of the constructor

File: keyvi/include/keyvi/dictionary/fuzzy_match.h [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+ public:
+  FuzzyMatch(fsa::automata_t fsa, const std::string& query, size_t max_edit_distance)
+      : fsa_(std::move(fsa)), query_(query), max_edit_distance_(max_edit_distance) {
+    std::vector<int> codepoints;
+    utf8::unchecked::utf8to32(query.begin(), query.end(), back_inserter(codepoints));
+
+    query_char_length_ = codepoints.size();
+    metric_ptr_.reset(new stringdistance::Levenshtein(codepoints, 20, max_edit_distance_));
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/fuzzy_match.h]: I suggest to move that down after the prefix has been matched, so in case prefix does not match we avoid initialization cost

File: keyvi/include/keyvi/dictionary/fuzzy_match.h [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+  fsa::automata_t fsa_;
+  std::string query_;
+  const size_t max_edit_distance_;
+
+  size_t exact_prefix_in_chars_ = 0;
+  size_t query_char_length_ = 0;
+
+  Match first_match_;
+
+  std::unique_ptr<stringdistance::Levenshtein> metric_ptr_;
```
Comment (hendrikmuhs) [inline keyvi/include/keyvi/dictionary/fuzzy_match.h]: that looks quite nice, so it would be easy to turn that into 

template<class MetricT = stringdistance::Levenshtein>
class FuzzyMatch final { 
...
};

Not saying we need that now.

File: keyvi/include/keyvi/dictionary/dictionary.h (hunk 3 above)
Comment (hendrikmuhs) [reply keyvi/include/keyvi/dictionary/dictionary.h]: what about fuzzy_matcher.h placed in a 'matching' subfolder?

In the long run I would migrate e.g. GetNear there as well. Yours look cleaner than my in-class implementations.

File: keyvi/tests/keyvi/dictionary/fuzzy_match_test.cpp [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+#include <boost/test/unit_test.hpp>
+
+#include "dictionary/dictionary.h"
+#include "testing/temp_dictionary.h"
+
+namespace keyvi {
+namespace dictionary {
+
+BOOST_AUTO_TEST_SUITE(FuzzyMatchTests)
```
Comment (hendrikmuhs) [inline keyvi/tests/keyvi/dictionary/fuzzy_match_test.cpp]: could you add some test cases for edge cases:

- no match
- empty string
- query shorter than prefix length

some CJK would be nice as well, although you use umlauts.

Comment (hendrikmuhs) [reply keyvi/tests/keyvi/dictionary/fuzzy_match_test.cpp]: some tests for the score would be nice as well. E.g. when it comes to non-ascii it is important that e.g. a mismatch of 'ü' scores edit distance 1 (codepoint length) and not 2 (byte length).

**Comment by hendrikmuhs** [comment]
added some comments. Looks pretty good!

File: keyvi/tests/keyvi/dictionary/fuzzy_match_test.cpp (hunk 9 above)
Comment (narekgharibyan) [reply keyvi/tests/keyvi/dictionary/fuzzy_match_test.cpp]: Sure!

File: keyvi/include/keyvi/dictionary/dictionary.h (hunk 4 above)
Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/dictionary.h]: Yup!

File: keyvi/include/keyvi/dictionary/dictionary.h (hunk 3 above)
Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/dictionary.h]: Yes, truth be told I was not sure about naming and location either. But I wanted to have it outside of Dictionary class for sure!
Also during testing I've realized that current implementation of `FuzzyCompletion` has a couple of bugs, and as it shares most of code with `FuzzyMatch`, later on it can be reimplemented by just deriving from `FuzzyMatch` and overriding `NextMatch()` method.

File: keyvi/include/keyvi/dictionary/fuzzy_match.h (hunk 8 above)
Comment (narekgharibyan) [reply keyvi/include/keyvi/dictionary/fuzzy_match.h]: Sure thing!

File: keyvi/include/keyvi/dictionary/dictionary.h (hunk 3 above)
Comment (hendrikmuhs) [reply keyvi/include/keyvi/dictionary/dictionary.h]: that sounds like a good idea, but again we have to differ between the matching and the completion case, still a lot of code can be shared.

FWIW 

 - fuzzy completion/matching isn't well tested, I agree
 - it has been written before I rewrote lots of the traversal code (https://github.com/KeyviDev/keyvi/blob/master/keyvi/include/keyvi/dictionary/fsa/traversal/traversal_base.h), which can make a big difference in speed
 - bugs: sure, please open issues for them, I really like to improve the quality of fuzzy matching
